stanford-earth/stanford_earth_helper
# Earth 913 Tag Events by Feed URL and Department/Program & fix More Info link

**READY FOR REVIEW**

**Summary**

Need to be able to track which feed URL an imported event comes from and tag it with the appropriate department or program based on that feed. Problem is that when the migration source configuration includes a list of many feed URLs, migrate_plus provides no obvious way to match an event's migration input row to the URL from which it came. Solved the problem by extending several migrate_plus classes and overriding the Drupal\migrate_plus\Plugin\migrate\source\Url:next function. This works, but was this the best approach? Would it have been simpler to just write patches for those classes? I can always go back and do that, I suppose.

A summary of what I did:
1. Add new stanford_earth_helper sub-module stanford_earth_migrate_extend
  1.1 Class Drupal\stanford_earth_migrate_extend\Plugin\migrate_plus\data_parser\StanfordEarthSimpleXml extends Drupal\migrate_plus\Plugin\migrate_plus\data_parser\SimpleXml with getter function for protected $activeUrl property.
  1.2 Class Drupal\stanford_earth_migrate_extend\Plugin\migrate_plus\data_parser\StanfordEarthJson extends Drupal\migrate_plus\Plugin\migrate_plus\data_parser\Json with getter function for protected $activeUrl property. (not needed by events but will be for CAP).
  1.3 Class Drupal\stanford_earth_migrate_extend\Plugin\migrate\source\StanfordEarthUrl extends Drupal\migrate_plus\Plugin\migrate\source\Url, overridding its next() function to store the current migration row's feed URL in the $row_data array as 'current_feed_url'.
  1.4  Class Drupal\stanford_earth_migrate_extend\Plugin\migrate\process\StanfordEarthFeedUrl is a migration process plugin which simply returns the 'current_feed_url' from the migration row to be further processed by the migrate_lookup plugin.
  1.5 Class Drupal\stanford_earth_migrate_extend\Plugin\migrate\process\StanfordEarthFeedDept is a migration process plugin which gets the 'current_feed_url' from the migration row, finds its matching taxonomy term, and uses the term's description to attempt to return the appropriate department or program name. The migrate_lookup plugin is then used to add the matching department/program taxonomy term to the event entity.

2. The form Drupal\stanford_earth_events\Form\EventImportersForm has been updated so that its submit hook checks if a taxonomy term exists for each specified URL. If not, the feed is requested and a term is created with the URL as the term name and the feed's title attribute used as the term's description.

3. The migration configurations migration_plus.migration.events_importer.yml and migration_plus.migration.events_importer_unlisted.yml have been updated to use the new source and process plugins defined above. Also changes the field_s_event_more_info source selector from 'link' to 'url' to get the proper link.

See also matching pull request for SE3_BLT which has additional configuration updates (taxonomy vocabularies, event content type fields, etc.) and instructions for testing.

**Needed By:**
End of next sprint

**Urgency**
The usual - needed to continue creating department/program event views.

**Steps to Test**
See the EARTH-913 pull request -- https://github.com/stanford-earth/SE3_BLT/pull/138 -- for se3_blt for testing steps.
